### PR TITLE
Pass inputs to metrics

### PR DESCRIPTION
**Pull request by yoavkatz**
Closes https://github.ibm.com/IBM-Research-AI/fm-eval/issues/671

This PR deals with the need to pass other values to metrics beside prediction and references.  This is important for example, for reference less metrics that want to compare the prediction to some context in the tasks (for example, entailment between context and predicted answer in an NLI task).

### CODE REVIEW

File: prepare/metrics/rag.py [hunk 1]
```diff
@@ -10,6 +19,58 @@
...
+predictions = ["apple", "boy", "cat"]
+references = [["apple2"], ["boys"], ["dogs"]]
+additional_inputs = [{"context": "apple 2e"}, {"context": "boy"}, {"context": "dog"}]
+instance_targets = [  # nDCG is undefined at instance level
+    {"f1": 0.67, "precision": 1.0, "recall": 0.5, "score": 0.67, "score_name": "f1"},
+    {"f1": 1.0, "precision": 1.0, "recall": 1.0, "score": 1.0, "score_name": "f1"},
+    {"f1": 0, "precision": 0, "recall": 0, "score": 0, "score_name": "f1"},
+]
+
+#   Currently rename fields does not delete the current fields
```
Comment (matanor) [inline prepare/metrics/rag.py:34]: remove comment?

File: prepare/metrics/rag.py [hunk 2]
```diff
@@ -10,6 +19,58 @@
...
+    "score_name": "f1",
+}
+
+metric = MetricPipeline(
+    main_score="score",
+    preprocess_steps=[
+        CopyFields(field_to_field=[("additional_inputs/context", "references")], use_query=True),
+        ListFieldValues(fields=["references"], to_field="references"),
+    ],
+    # metric=SentenceBert(model_name="sentence-transformers/all-mpnet-base-v2"),
```
Comment (matanor) [inline prepare/metrics/rag.py:53]: remove commented out line?

File: src/unitxt/blocks.py [hunk 3]
```diff
@@ -22,6 +22,7 @@
 from .processors import ToString, ToStringStripped
 from .recipe import SequentialRecipe
 from .splitters import RandomSampler, SliceSplit, SplitRandomMix, SpreadSplit
+from .standard import StandardRecipe
```
Comment (matanor) [inline src/unitxt/blocks.py:25]: why is this added? (only this line changed in the file?)

File: src/unitxt/dataset.py [hunk 4]
```diff
@@ -91,6 +91,7 @@ def get_dataset_artifact(dataset_str):
         if "type" not in args:
             args["type"] = os.environ.get("UNITXT_DEFAULT_RECIPE", __default_recipe__)
         recipe = Artifact.from_dict(args)
+    print(recipe)
```
Comment (matanor) [inline src/unitxt/dataset.py:94]: remove print

File: src/unitxt/metric.py [hunk 5]
```diff
@@ -98,11 +99,16 @@ def process(self, predictions: List[str], references: Iterable, split_name: str
 from .schema import UNITXT_DATASET_SCHEMA
 
 
+def _from_key_value_pairs(key_value_list: Dict[str, list]) -> Dict[str, str]:
```
Comment (matanor) [inline src/unitxt/metric.py:102]: why is the Dict from str to a list? This is the general structure of the input?

File: src/unitxt/metric.py [hunk 6]
```diff
@@ -98,11 +99,16 @@ def process(self, predictions: List[str], references: Iterable, split_name: str
...
+    return dict([(key, value) for key, value in zip(key_value_list["key"], key_value_list["value"])])
+
+
 class MetricRecipe(SequentialOperatorInitilizer):
     def prepare(self):
         register_all_artifacts()
         self.steps = [
             FromPredictionsAndOriginalData(),
+            Apply("additional_inputs", function=_from_key_value_pairs, to_field="additional_inputs"),
```
Comment (matanor) [inline src/unitxt/metric.py:111]: again, maybe document a bit what is going on. AFAIU this operator processes the "additional_inputs" field and converts it from the two keyed-dictionary to a dict where the keys are.. the input and output fields?

File: src/unitxt/metrics.py [hunk 7]
```diff
@@ -58,10 +67,22 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
...
-            refs, pred = instance["references"], instance["prediction"]
-
+            instance_references, instance_prediction = instance["references"], instance["prediction"]
+            references.append(instance_references)
+            predictions.append(instance_prediction)
+            instances.append(instance)
+            if self.use_additional_inputs:
+                assert (
+                    "additional_inputs" in instance
+                ), f"'additional_inputs' field expected in all instances of passed to metric {self.__class__.__name__}. Provided instance: {instance}"
```
Comment (matanor) [inline src/unitxt/metrics.py:77]:
```suggestion
                ), f"'additional_inputs' field expected in all instances passed to metric {self.__class__.__name__}. Provided instance: {instance}"
```

File: src/unitxt/metrics.py [hunk 8]
```diff
@@ -70,26 +91,40 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
...
-        result = self.compute(references, predictions)
+    def _compute(self, references: List[List[str]], predictions: List[str], additional_inputs: List[Any]) -> dict:
+        result = self.compute_with_additional_inputs(references, predictions, additional_inputs)
         result["score"] = result[self.main_score]
         result["score_name"] = self.main_score
         return result
 
     @abstractmethod
-    def compute(self, references: List[List[str]], predictions: List[str]) -> dict:
+    def compute_with_additional_inputs(
```
Comment (matanor) [inline src/unitxt/metrics.py:109]: why was this function renamed? given that there is not regular `compute`, why not just keep it `compute`?

File: src/unitxt/metrics.py [hunk 9]
```diff
@@ -38,16 +38,25 @@ def process(self, instance: Dict[str, Any], stream_name: str = None) -> Dict[str
 
 # TODO: currently we have two classes with this name. metric.Metric and matrics.Metric...
 class Metric(ABC):
+    use_additional_inputs: bool = True
```
Comment (matanor) [inline src/unitxt/metrics.py:41]: what is the downside of always providing the additional inputs? in other words, while support setting this flag to False?

File: src/unitxt/metrics.py [hunk 10]
```diff
@@ -191,24 +259,32 @@ def process(self, stream: Stream, stream_name: str = None) -> Generator:
...
+        super().prepare()
+        self.use_additional_inputs = False
+
+    def compute_with_additional_inputs(self, references: List[Any], prediction: Any, additional_inputs: Dict) -> dict:
+        assert additional_inputs is None
+        return self.compute(references, prediction)
 
     @abstractmethod
-    def compute(self, references: List[str], prediction: str) -> dict:
+    def compute(self, references: List[Any], prediction: Any, additional_inputs: Dict) -> dict:
```
Comment (matanor) [inline src/unitxt/metrics.py:277]: why pass `additional_inputs` here? isn't `compute_with_additional_inputs` for that?

File: src/unitxt/metrics.py [hunk 11]
```diff
@@ -224,24 +300,11 @@ def compute(self, references: List[List[str]], predictions: List[str]) -> dict:
...
-        pass
-
-
-class Accuracy(InstanceMetric):
+class Accuracy(ReferenceBasedInstanceMetric):
     reduction_map = {"mean": ["accuracy"]}
     main_score = "accuracy"
 
-    def compute(self, references: List[str], prediction: str) -> dict:
+    def compute(self, references: List[Any], prediction: Any) -> dict:
```
Comment (matanor) [inline src/unitxt/metrics.py:307]: why were the type hints changed?

File: src/unitxt/metrics.py [hunk 12]
```diff
@@ -291,7 +354,7 @@ def prepare(self):
         super().prepare()
         self.metric = evaluate.load(self.hf_metric_name, experiment_id=self.experiment_id)
 
-    def compute(self, references: List[List[str]], predictions: List[str]) -> dict:
+    def compute(self, references: List[List[Any]], predictions: List[Any]) -> dict:
```
Comment (matanor) [inline src/unitxt/metrics.py:357]: same q: why were the type hints changed?

File: src/unitxt/metrics.py [hunk 13]
```diff
@@ -496,9 +559,10 @@ class Rouge(HuggingfaceMetric):
     sent_split_newline: bool = True
 
     def prepare(self):
+        super().prepare()
```
Comment (matanor) [inline src/unitxt/metrics.py:562]: why was this changed? just checking it ok

File: src/unitxt/metrics.py [hunk 14]
```diff
@@ -511,22 +575,25 @@ def compute(self, references, predictions):
         return super().compute(references, predictions)
 
 
-# Computes char edit distance, ignoring whitespace
-class CharEditDistanceAccuracy(SingleReferenceInstanceMetric):
+# Computes chat edit distance, ignoring whitespace
```
Comment (matanor) [inline src/unitxt/metrics.py:578]: chat or char? asking because the class remains char..

File: src/unitxt/metrics.py [hunk 15]
```diff
@@ -804,15 +872,55 @@ def compute(self, references: List[List[Any]], predictions: List[Any]) -> List[A
...
     def prepare(self):
         from sklearn.metrics import ndcg_score
 
+        super().prepare()
         self.eval = ndcg_score
 
-    def compute(self, references: List[List[float]], predictions: List[float]) -> dict:
-        scores_dict = {self.main_score: self.eval([references], [predictions])}
+    def compute_with_additional_inputs(
+        self, references: List[List[Any]], predictions: List[Any], additional_inputs: List[Any]
```
Comment (matanor) [inline src/unitxt/metrics.py:900]: i am not going over this one in depth, i suppose it was written by both you and Ariel, hence i consider it well-reviewed :)

File: src/unitxt/schema.py [hunk 16]
```diff
@@ -32,7 +33,13 @@ class ToUnitxtGroup(StreamInstanceOperatorValidator):
     postprocessors: List[str] = field(default_factory=lambda: ["to_string_stripped"])
     remove_unnecessary_fields: bool = True
 
+    def _to_lists_of_keys_and_values(self, dict: Dict[str, str]):
+        return {"key": [key for key, _ in dict.items()], "value": [str(value) for _, value in dict.items()]}
```
Comment (matanor) [inline src/unitxt/schema.py:37]:
```suggestion
        return {"key": list(dict.keys()), "value": [str(value) for value in dict.values()]}
```
or does that make an issue with the iteration order, that is not guaranteed to be the same for `keys` or `values`?

File: tests/test_examples.py [hunk 17]
```diff
@@ -55,7 +79,7 @@ def test_add_metric_to_catalog(self):
 
         metric = evaluate.load(unitxt.metric_file)
 
-        results = metric.compute(predictions=predictions, references=data, flatten=True)
+        results = metric.compute(predictions=predictions, references=data, flatten=False)
```
Comment (matanor) [inline tests/test_examples.py:82]: why has this changed?

File: tests/test_examples.py [hunk 18]
```diff
@@ -132,15 +180,14 @@ def test_example7(self):
 
     def test_example8(self):
         dataset = unitxt.load_dataset("recipes.wnli_3_shot")
-
+        print(dataset["test"][0])
```
Comment (matanor) [inline tests/test_examples.py:183]: remove print?

**Review by matanor — CHANGES_REQUESTED** [review]
hi @yoavkatz , left some comments, most are minor.
One major question i had is why keep the option of not passing the additional inputs to the metrics? with the flag that was added to Metric?

File: prepare/metrics/ndcg.py [hunk 19]
```diff
@@ -10,29 +10,33 @@
         CopyFields(field_to_field=[("references/0", "references")], use_query=True),
         CastFields(
             fields={"prediction": "float", "references": "float"},
-            failure_defaults={"prediction": 0.0},
+            failure_defaults={"prediction": None},
```
[the remark this answers is not in the document]
Comment (yoavkatz) [reply prepare/metrics/ndcg.py:13]: It a fix by Ariel - to ensure that if predictions are can not be casted to float, they are handled in a special way in the metrics, and not as a 0 score, which has semantic meaning in ordering the predictions.

File: examples/example_mixture.py [hunk 20]
```diff
@@ -1,6 +1,7 @@
 from src.unitxt import add_to_catalog, load_dataset
 from src.unitxt.blocks import LoadHF
 from src.unitxt.fusion import WeightedFusion
+from src.unitxt.standard import StandardRecipe
```
[the remark this answers is not in the document]
Comment (yoavkatz) [reply examples/example_mixture.py:4]: Removed.

File: prepare/metrics/rag.py [hunk 21]
```diff
@@ -10,6 +19,58 @@
...
     ("metrics.reward.deberta.v3.large.v2", Reward(model_name="OpenAssistant/reward-model-deberta-v3-large-v2")),
 ]
-
 for metric_id, metric in metrics:
     add_to_catalog(metric, metric_id, overwrite=True)
+
+predictions = ["apple", "boy", "cat"]
+references = [["apple2"], ["boys"], ["dogs"]]
+additional_inputs = [{"context": "apple 2e"}, {"context": "boy"}, {"context": "dog"}]
+instance_targets = [  # nDCG is undefined at instance level
```
[the remark this answers is not in the document]
Comment (yoavkatz) [reply prepare/metrics/rag.py:28]: Done

File: prepare/metrics/rag.py (hunk 1 above)
Comment (yoavkatz) [reply prepare/metrics/rag.py:34]: Clarified comment.

File: prepare/metrics/rag.py (hunk 2 above)
Comment (yoavkatz) [reply prepare/metrics/rag.py:53]: Done.

File: src/unitxt/blocks.py (hunk 3 above)
Comment (yoavkatz) [reply src/unitxt/blocks.py:25]: Deleted.

File: src/unitxt/dataset.py (hunk 4 above)
Comment (yoavkatz) [reply src/unitxt/dataset.py:94]: Deleted.

File: src/unitxt/metric.py (hunk 6 above)
Comment (yoavkatz) [reply src/unitxt/metric.py:111]: Added documentation.

Regarding the second questions, . It is the field from which the value the Apply function receives is called.

File: src/unitxt/metrics.py (hunk 7 above)
Comment (yoavkatz) [reply src/unitxt/metrics.py:77]: Fixed.

File: src/unitxt/metrics.py (hunk 9 above)
Comment (yoavkatz) [reply src/unitxt/metrics.py:41]: Worth considering.  Let's discuss.

File: src/unitxt/metrics.py (hunk 11 above)
Comment (yoavkatz) [reply src/unitxt/metrics.py:307]: This is the current behavior.  The processed predictions and references are not strings.

File: src/unitxt/metrics.py (hunk 10 above)
Comment (yoavkatz) [reply src/unitxt/metrics.py:277]: Fixed.

File: src/unitxt/metrics.py (hunk 12 above)
Comment (yoavkatz) [reply src/unitxt/metrics.py:357]: Same as before.  processed predictions  and references are not always string. (e.g. in NER)

File: src/unitxt/metrics.py (hunk 14 above)
Comment (yoavkatz) [reply src/unitxt/metrics.py:578]: Fixed.

File: src/unitxt/metrics.py (hunk 15 above)
Comment (yoavkatz) [reply src/unitxt/metrics.py:900]: Sure...

File: src/unitxt/schema.py (hunk 16 above)
Comment (yoavkatz) [reply src/unitxt/schema.py:37]: I'm not sure.  Just wanted to keep it simple and consistent.

File: tests/test_examples.py [hunk 22]
```diff
@@ -105,7 +153,7 @@ def test_example7(self):
 
         metric = evaluate.load(unitxt.metric_file)
 
-        results = metric.compute(predictions=predictions, references=data, flatten=True)
+        results = metric.compute(predictions=predictions, references=data, flatten=False)
```
[the remark this answers is not in the document]
Comment (yoavkatz) [reply tests/test_examples.py:156]: Fixed.

File: prepare/metrics/rag.py [hunk 23]
```diff
@@ -10,6 +19,61 @@
...
+predictions = ["apple", "boy", "cat"]
+references = [["apple2"], ["boys"], ["dogs"]]
+additional_inputs = [{"context": "apple 2e"}, {"context": "boy"}, {"context": "dog"}]
+instance_targets = [
+    {"f1": 0.67, "precision": 1.0, "recall": 0.5, "score": 0.67, "score_name": "f1"},
+    {"f1": 1.0, "precision": 1.0, "recall": 1.0, "score": 1.0, "score_name": "f1"},
+    {"f1": 0, "precision": 0, "recall": 0, "score": 0, "score_name": "f1"},
+]
+
+# Currently rename fields does not delete the current fields,
```
Comment (matanor) [inline prepare/metrics/rag.py:34]:
```suggestion
# Currently CopyFields does not delete the source fields,
```

File: examples/add_metric_to_catalog.py [hunk 24]
```diff
@@ -15,6 +15,7 @@
 ]
 
 for d in data:
+    d["additonal_inputs"] = []
```
Comment (matanor) [inline examples/add_metric_to_catalog.py:18]: shouldn't this be {}?

**Review by matanor — APPROVED** [review]
LGTM. left one small comment
